---
title: "larvae_paper"
output: html_document
---

```{r setup, include=FALSE}
pkg = c("tidyverse", 
        "rnaturalearth", 
        "rnaturalearthdata", 
        "sf", 
        "cowplot", 
        "parallel", 
        "gridExtra", 
        "grid", 
        "data.table",
        'phangorn', 
        'phytools', 
        'pegas',
        'vcfR', 
        'RColorBrewer', 
        'rehh')

#install.packages(pkg) #install packages if you need them and load
new.packages <- pkg[!(pkg %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)
lapply(pkg, require, character.only = TRUE)#
knitr::opts_chunk$set(echo = TRUE)
metadata=fread('~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/metadata/metadata_full_sequencedonly_062022.csv')
metadata$Species = ifelse(metadata$Form=='M', 'A. coluzzi', 'A. gambiae')
```



```{r fig1}
#load geog objects
world <- ne_countries(scale = 'large', returnclass = "sf", continent = 'africa')
lakes <- ne_download(scale = 'large', type = 'lakes', category = 'physical', returnclass = "sf")
rivers <- ne_download(scale = 'large', type = 'rivers_lake_centerlines', category = 'physical', returnclass = "sf")
ocean <- ne_download(scale = 'large', type = 'ocean', category = 'physical', returnclass = "sf")
#get site only data and coilour by habitat (darken for border)
sites = metadata %>% select(Site, Lat, Long, habitat) %>% unique() #%>% full_join(colesites)
habitatocolours = c('#e41a1c','#377eb8','#4daf4a','#984ea3')
darkerhabitatcolours3 = colorspace::darken(habitatocolours, amount=0.3)
#plot sites coloured by isec and eco, and species 
maps = ggplot()+
  geom_sf(data=world, color = '#dfc27d', fill='#f6e8c3')+
  geom_sf(data = rivers, colour = '#4575b4', fill = '#a6bddb')+
  geom_sf(data = lakes, colour = '#4575b4', fill = '#a6bddb')+
  geom_sf(data = ocean, colour = '#4575b4', fill = '#a6bddb')+
  geom_point(data=metadata, aes(y=Lat,x=Long, color=habitat, fill=habitat, shape=INSECTICIDE_USE), size = 6)+
  scale_color_manual(values=darkerhabitatcolours3,guide='none')+
  scale_fill_manual(values=habitatocolours)+
  scale_shape_manual(values=c(21, 22, 23, 24))+
  ggspatial::annotation_scale(location = "br", width_hint = 0.4) +
  coord_sf(xlim = c(-3.6, 1.5), ylim = c(4.5, 11.5), expand = FALSE)+
  labs(x='Longitude', y='Latitude', color='Ecozone',fill = 'Ecozone', shape='Insecticide Usage Pattern')+
  guides(fill=guide_legend(override.aes=list(colour=habitatocolours)))+
  theme_classic()+
  facet_wrap(~Species)
#
ggsave(plot=maps,filename = '~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/figures/fig1_maps.tiff', device = grDevices::tiff, width = 30, height = 15, units = 'cm', dpi = 1000)
```

```{r fig2}
#--------------------------------------------------#
#LD decay                                          #
#--------------------------------------------------#

#read decay file
decay = read.csv('~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/data/structure/ld_decay_data_bychr.csv')
#sub chrnames for more readable versions
decay$Scaffold = str_replace(decay$File,'_glf2_minorfilts_angsd_samtools_ngsLD', '')
decay$Scaffold = str_replace(decay$Scaffold,'AgamP4_', '')

#make ld decay plot
ld_decay = decay %>% 
  ggplot(aes(x=Dist/1000, y =value, color=Scaffold))+
  geom_line()+
  facet_wrap(~Scaffold)+
  theme_classic()+
  labs(x='\nDistance (kb)', y='LD (r2)', colour = 'Scaffold')+
  theme(axis.text.x = element_text(angle = 30, vjust = 0.5, hjust=1))+
  theme(legend.position = "none",
        axis.title=element_text(size=12),
        axis.title.x = element_text(margin = margin(t = 0, r = 0, b = 15, l = 0)))

#--------------------------------------------------#
#pca                                               #
#--------------------------------------------------#

x = fread('~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/data/structure/3_euch_pruned.cov')
s = eigen(x)
f = cbind(s$vectors[,1:2], seq(0, length(s$vectors[,1])-1))
t= as.data.frame(f) %>% left_join(., metadata, by = c('V3' = 'bamorder'))

#how much variation can be explained by the PCs?
v = s$values/sum(s$values)

v = cbind(as.data.frame(v), seq(1, length(s$vectors[,1])))

colnames(v) <- c('var','pc')

v$pcs = paste0('PC', v$`seq(1, length(s$vectors[, 1]))`)

integer_breaks <- function(n = 10, ...) {
  fxn <- function(x) {
    breaks <- floor(pretty(x, n, ...))
    names(breaks) <- attr(breaks, "labels")
    breaks
  }
  return(fxn)
}

v = v %>% filter(pc < 11)

#pca variation explained plot
pca_var_plot = 
  ggplot(v, aes(y=var*100, x=pc))+
  geom_line()+
  theme_minimal()+
  scale_x_continuous(breaks=integer_breaks())+
  labs(x='PCs', y='% Variation Explained by PC')

#make pca plot
pcaplot = t %>% 
  ggplot(aes(x=V1, y=V2, color=Species))+
  geom_point()+
  theme_classic()+
  labs(x='PC1', y='PC2')+  
  scale_color_brewer(palette = 'Dark2')+
  theme(axis.title = element_text(size = 12))

#--------------------------------------------------#
#admixture                                         #
#--------------------------------------------------#
#load pcangsd admixture output
q = fread('~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/data/structure/3_euch_pruned.Q')

#make a df of population assignment proportion for each individual (needs to join to metadata by bam order)
admixture_plot_data = q %>% mutate(id = row_number()) %>% pivot_longer(cols=c(V1, V2, V3, V4, V5), names_to = 'pop', values_to = 'prob') %>% left_join(., metadata, by = c('id' = 'bamorder')) %>% select(pop, prob, id, species) %>% 
  group_by(id) %>% 
  mutate(main_proportion = pop[which.max(prob)], assignment_prob = max(prob)) %>% 
  arrange(main_proportion, desc(assignment_prob)) %>%  
  ungroup() %>% 
  mutate(id = forcats::fct_inorder(factor(id)))

admixchah = ggplot(admixture_plot_data, aes(id, prob, fill = pop)) +
  geom_bar(position="fill", stat="identity", width = 1) +
  theme_minimal() +
  labs(x='Individual', y= 'Admixture Proportion', fill = 'Cluster')+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+
  scale_fill_brewer(palette = 'Dark2')+
  theme(
    plot.margin = margin(0.5, 0.5, 0.5, 0.5, unit = "cm"),
    axis.text.x=element_blank(),
    axis.ticks.x=element_blank(),
    axis.title = element_text(size = 12),
    plot.title = element_text(size = 15, face = "bold"),
    strip.text.y = element_text(angle = 270, face = "bold"),
    strip.placement = "outside",
    panel.grid.major.y = element_blank(),
    strip.text.x = element_blank())+
  facet_grid(~main_proportion, scales = 'free_x', space = "free_x")

grid1 = cowplot::plot_grid(ld_decay, pcaplot,labels = c('A', 'B'), align = "h", axis = 'l', label_size = 15)
fig2grid = plot_grid(grid1, admixchah, ncol=1, labels = c('','C'), label_size = 15)
ggsave(plot=fig2grid,filename = '~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/figures/fig2grid.tiff', device = grDevices::tiff, width = 25, height = 15, units = 'cm', dpi = 1000)
```
```{r fig3}
#make chr colour palette
colpal = c('AgamP4_2L', 'AgamP4_2R', 'AgamP4_3L', 'AgamP4_3R', 'AgamP4_X')
colpal = RColorBrewer::brewer.pal(5, "Set2")

#function to turn angsd genome scan output into something plottable
make_pbs_df <- function(pbsdffile) {
  pbs_df = fread(pbsdffile)
  colnames(pbs_df) <- c('Angsd_gubbins', 'chr', 'pos', 'nsites', 'fst01', 'fst02', 'fst12', 'PBS0', 'PBS1', 'PBS2')
  pbs_df = pbs_df[grep("Agam", chr)]
  pbs_df = pbs_df[!grep("UNK", chr)]
  pbs_df = pbs_df[!grep("unplaced", chr)]
  chrlens = aggregate(pbs_df$pos, by = list(pbs_df$chr), max) #get chromosome lengths (mx bp)
  colnames(chrlens) = c('chr', 'chrlen') #rename
  chrlens$cumsum = cumsum(chrlens$chrlen) - chrlens$chrlen #calculate cumulative positions of each chr start
  pbs_df = pbs_df[chrlens, on = .(chr)] #join to thetas data
  pbs_df$BPcum = pbs_df$pos + pbs_df$cumsum #get cumulative position in genome
  #pbs_df <- sample_n(pbs_df, size = nrow(pbs_df) * 0.5)
  return(pbs_df)
}

#an col (fig 3)
mform_mh <- list.files('/Users/tristanpwdennis/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/data/genome_scans/insecticide_usage/by_ecotype_and_high_isecusage_mangrove/', full.names = T)
#make list of dfs
hilist <- mclapply(mform_mh, make_pbs_df, mc.cores = 6)
#chr color palette
pal = c('#d73027', '#fc8d59', '#91bfdb', '#4575b4', '#fee090')

#df: mangrove high
ac_df_mh = ggplot(hilist[[1]], aes(x=BPcum, y=fst02, colour = chr))+
  geom_line()+
  ylim(0, 1)+
  theme_classic()+
  scale_color_manual(values = pal)+
  theme(legend.position = "none", axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank(), axis.title.y=element_blank(),
        plot.margin = margin(0,0,0,1, "cm"))+  labs(x='Position', y='Fst')+
      ggtitle('Deciduous Forest: Mangrove, High IS usage')

#rf: mangrove high
ac_rf_mh = ggplot(hilist[[1]], aes(x=BPcum, y=fst12, colour = chr))+
  geom_line()+
  ylim(0, 1)+
  theme_classic()+
  scale_color_manual(values = pal)+
  theme(legend.position = "none", axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank(), axis.title.y=element_blank(),
        plot.margin = margin(0,0,0,1, "cm"))+
  labs(y='Fst')+
    ggtitle('Rainforest: Mangrove, High IS usage')

#rf: mangrove high
ac_cs_mh = ggplot(hilist[[2]], aes(x=BPcum, y=fst12, colour = chr))+
  geom_line()+
  ylim(0, 1)+
  theme_classic()+
  scale_color_manual(values = pal)+
  labs(y='Position')+
  theme(legend.position = "none", axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank(), axis.title.y=element_blank(),
        plot.margin = margin(0,0,0,1, "cm"))+  labs(x='Position', y='Fst')+
  ggtitle('Coastal Savannah: Mangrove, High IS usage')

#common x and y labs
#common x and y labs
y.grob <- textGrob("Fst", 
                   gp=gpar(fontface="bold", col="black", fontsize=10), rot=90, vjust = 1.8)

x.grob <- textGrob("Position", 
                   gp=gpar(fontface="bold", col="black", fontsize=10))

ac_fig <- plot_grid(ac_df_mh, ac_rf_mh, ac_cs_mh, ncol =1, align = 'vh')
ac_fig <- grid.arrange(arrangeGrob(ac_fig, left = y.grob, bottom = x.grob))

#ac_fig = egg::ggarrange(ac_df_mh, ac_rf_mh, ac_cs_mh, ncol = 1, nrow = 3, labels=c('A', 'B', 'C'))
ggsave(plot=ac_fig,filename = '~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/figures/fig3_ancol_fst.tiff', device = grDevices::tiff, width = 20, height = 10, units = 'cm')
```
```{r fig4}
#make chr colour palette
colpal = c('AgamP4_2L', 'AgamP4_2R', 'AgamP4_3L', 'AgamP4_3R', 'AgamP4_X')
colpal = RColorBrewer::brewer.pal(5, "Set2")

#function to turn angsd genome scan output into something plottable
make_pbs_df <- function(pbsdffile) {
  pbs_df = fread(pbsdffile)
  colnames(pbs_df) <- c('Angsd_gubbins', 'chr', 'pos', 'nsites', 'fst01', 'fst02', 'fst12', 'PBS0', 'PBS1', 'PBS2')
  pbs_df = pbs_df[grep("Agam", chr)]
  pbs_df = pbs_df[!grep("UNK", chr)]
  pbs_df = pbs_df[!grep("unplaced", chr)]
  chrlens = aggregate(pbs_df$pos, by = list(pbs_df$chr), max) #get chromosome lengths (mx bp)
  colnames(chrlens) = c('chr', 'chrlen') #rename
  chrlens$cumsum = cumsum(chrlens$chrlen) - chrlens$chrlen #calculate cumulative positions of each chr start
  pbs_df = pbs_df[chrlens, on = .(chr)] #join to thetas data
  pbs_df$BPcum = pbs_df$pos + pbs_df$cumsum #get cumulative position in genome
  #pbs_df <- sample_n(pbs_df, size = nrow(pbs_df) * 0.5)
  return(pbs_df)
}

#angam (figure)
mform_mh <- list.files('/Users/tristanpwdennis/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/data/genome_scans/insecticide_usage/by_ecotype_and_low_isecusage_mangrove/', full.names = T)
hilist <- mclapply(mform_mh[[7]], make_pbs_df, mc.cores = 6)

#df: rf and dfl
ag_rf_dfl = ggplot(hilist[[1]], aes(x=BPcum, y=fst02, colour = chr))+
  geom_line()+
  ylim(0, 1)+
  theme_classic()+
  scale_color_manual(values = pal)+
  theme(legend.position = "none", axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank(), axis.title.y=element_blank(),
        plot.margin = margin(0,0,0,1, "cm"))+  labs(x='Position', y='Fst')+
    ggtitle('Rainforest: Deciduous Forest, Low IS usage')

#rf: cs and dfl
ag_cs_dfl = ggplot(hilist[[1]], aes(x=BPcum, y=fst12, colour = chr))+
  geom_line()+
  ylim(0, 1)+
  theme_classic()+
  scale_color_manual(values = pal)+
  theme(legend.position = "none", axis.title.x=element_blank(),axis.text.x=element_blank(),axis.ticks.x=element_blank(), axis.title.y=element_blank(),
        plot.margin = margin(0,0,0,1, "cm"))+
  labs(x='Position', y='Fst')+
      ggtitle('Coastal Savannah: Deciduous Forest, Low IS usage')
#common x and y labs
y.grob <- textGrob("Fst", 
                   gp=gpar(fontface="bold", col="black", fontsize=10), rot=90, vjust = 1.8)

x.grob <- textGrob("Position", 
                   gp=gpar(fontface="bold", col="black", fontsize=10))


ag_fig <- plot_grid(ag_rf_dfl, ag_cs_dfl, ncol =1, align = 'hv')

ag_fig <- grid.arrange(arrangeGrob(ag_fig, left = y.grob, bottom = x.grob))

#ac_fig = egg::ggarrange(ac_df_mh, ac_rf_mh, ac_cs_mh, ncol = 1, nrow = 3, labels=c('A', 'B', 'C'))
ggsave(plot=ag_fig,filename = '~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/figures/fig5_angam_fst.tiff', device = grDevices::tiff, width = 20, height = 6.6, units = 'cm')
```

```{r cyp9ksleuthing}
#TREE make meta
#create haplotype specific metadata (duplicate ind rows and add hap1/hap2 designation)
meta_h1 = metadata
meta_h2 = metadata
meta_h1$hapid = 'h1'
meta_h2$hapid = 'h2'
hapmeta = rbind(meta_h1, meta_h2)
hapmeta$indid = paste0(hapmeta$seq_id, '_', hapmeta$hapid)

#load alignment, clean, add clean names
cypfas = read.FASTA('~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/data/cyp9k_hap_alignment/from_samtools/af_001.grouped.rn.fasta')
acefas = read.FASTA('~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/data/ace1_alignment/ace1_af_001.grouped.fasta')

names = gsub(pattern = "af_001/" , '',names(cypfas))
names = gsub(pattern = "geneonly" , '',names)
names(cypfas) = names
cypfas

#load phylogenies and make tip data
cyptreefile = '~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/data/cyp9k_hap_alignment/from_samtools/af_001.grouped.rn.fasta.treefile'
acetreefile = '~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/data/ace1_alignment/ace1_af_001.grouped.fasta.treefile'
treedata = dplyr::select(hapmeta, indid, hapid, habitat, Form, Site) #make tree data

plot_phy <- function(treefile) {
  phy = ape::read.tree(treefile)
  phy$tip.label <- gsub(pattern = 'geneonly', replacement = '', phy$tip.label)
  #read tree, root
  phy_p = midpoint.root(phy)
  #create ordered treedata
  treedata = treedata[match(phy$tip.label, treedata$indid), ] 
  #configure tip labels and set lengths and colours
  phy_p$species = treedata$Form 
  phy_p$site  =treedata$Site
  phy_p$habitat = treedata$habitat
  phy_p$tip.colors = c('#D95F02', "#1B9E77")
  phy_p$edge.length[phy_p$edge.length >  0.005] = 0.005
  phy_p$edge.length[phy_p$edge.length == 0]    = 5e-5
  spec <- factor(phy_p$species)
  mycol <- c('#D95F02', "#1B9E77")[spec]
  spec
  #pliot tree and tiplabs
  par(mar=c(0.5,0.5,0.5,0.5))
  plot.phylo(phy_p, type = 'unr',
             use.edge.length = T, 
             show.tip.label = F,
             lab4ut = "axial",
             edge.color = "slategray3", font=0.5, edge.width = 0.5, node.depth = 1)
  tiplabels(pch=16,col=mycol)
  add.scale.bar(cex=0.7, "topleft")
}

#make final phylogeny plot object
tiff(filename = '~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/figures/cyp9k1_phy.tiff')
plot_phy(cyptreefile)
title("cyp9k1                                                           ")
legend("bottomleft", legend = c("An. gambiae", "An. coluzzi"), pch = 22,
       pt.bg = c('#D95F02', "#1B9E77"), pt.cex = 2.5, bty="n", box.lty=1)
dev.off()





```


```{r selscan}
hstats = grep(pattern = 'X', (list.files(pattern = 'tsv', path = '/Users/tristanpwdennis/Projects/MOVE/bulky_larvae_paper_data/phased_gls_beagle4/', full.names = TRUE)), value = TRUE)
#read dfs and add filename as col
hstatdfs = lapply(hstats, function(i){data.frame(fread(i), name = i)})
hstatdf = do.call(rbind,hstatdfs)
hstatdf$stat = gsub(".tsv", "", gsub("mform_X_b4phasedgls_","",basename(hstatdf$name)))
hstatdf$stat =  gsub("sform_X_b4phasedgls_", "", hstatdf$stat)
hstatdf$form = gsub("\\_.*","",basename(hstatdf$name))

x = c(unique(hstatdf$stat))
doplot <- function(filterstat) {
  plotdf = filter(hstatdf, stat == filterstat)
  ggplot(plotdf, aes(x=midpoint/1000000, y=H2_H1, color=form))+
  geom_line()+
  scale_color_brewer(palette = 'Dark2')+
  facet_wrap(~form)+
  labs(y=filterstat)+
  theme_classic()+
  theme(strip.text.x = element_blank(), axis.title.x = element_blank(),legend.position = 'none', axis.text = element_text(size=7))
}

hstatplotlist = lapply(x, doplot)


mhapdiv = data.frame(t(fread('/Users/tristanpwdennis/Projects/MOVE/bulky_larvae_paper_data/phased_gls_beagle4/mform_X_hapdiv.txt')))
shapdiv = data.frame(t(fread('/Users/tristanpwdennis/Projects/MOVE/bulky_larvae_paper_data/phased_gls_beagle4/sform_X_hapdiv.txt')))


mhapdiv = cbind(mhapdiv, 'A. coluzzi','Haplotype Diversity')
shapdiv = cbind(shapdiv, 'A. gambiae','Haplotype Diversity')

colnames(mhapdiv) <- c('div', 'pos', 'spec', 'stat')
colnames(shapdiv) <- c('div', 'pos', 'spec', 'stat')
stathapdiv = rbind(mhapdiv, shapdiv)

divgrid = ggplot(stathapdiv, aes(x=pos/1000000, y=div,color= spec))+
  geom_line()+
  scale_color_brewer(palette = 'Dark2')+
  theme_classic()+
  facet_wrap(~spec)+
  theme(strip.text.x = element_blank(), axis.title.x = element_blank(),legend.position = 'none', axis.text = element_text(size=7))+
  labs(y='Hap Diversity')


v = plot_grid(divgrid, plotlist = hstatplotlist, ncol=1, align='hv')


ggsave(plot=v ,filename = '~/OneDrive - University of Glasgow/MOVE/manuscripts/td_je_angam_2022/figures/fig5_hstat.tiff', device = grDevices::tiff, width = 15, height = 20, units = 'cm')

```

